File: built-in plugins dsp/mb-vocal-debreath.cpp
/**
 * MB De-Breath
 * Category : effect
 * Type     : vocal
 * Version  : 1.0.0
 * Author   : Max Booster
 * Desc     : Breath noise reduction
 *
 * Part of Max Booster Built-In Plugins DSP
 */

#ifndef MB_VOCAL_DEBREATH_H
#define MB_VOCAL_DEBREATH_H

#include <algorithm>
#include <cmath>
#include <cstring>
#include "PluginBase.h"

class MbVocalDebreath : public PluginBase {
public:
    static constexpr const char* PLUGIN_ID      = "mb-vocal-debreath";
    static constexpr const char* PLUGIN_NAME    = "MB De-Breath";
    static constexpr const char* PLUGIN_TYPE    = "vocal";
    static constexpr const char* PLUGIN_CATEGORY = "effect";
    static constexpr const char* VERSION         = "1.0.0";

    struct Parameters {
    float sensitivity = 0.5f;  // range [0, 1]
    float reduction = -12f;  // range [-40, 0]
    };

    MbVocalDebreath() = default;
    ~MbVocalDebreath() override = default;

    void setSampleRate(double sampleRate) override {
        sampleRate_ = sampleRate;
        reset();
    }

    void reset() override {
        std::memset(buffer_, 0, sizeof(buffer_));
    }

    void process(float* left, float* right, int numSamples, Parameters params) {
        params.sensitivity = std::clamp(params.sensitivity, 0f, 1f);
        params.reduction = std::clamp(params.reduction, -40f, 0f);
        for (int i = 0; i < numSamples; ++i) {
            left[i]  = processSample(left[i],  params);
            right[i] = processSample(right[i], params);
        }
    }

private:
    double sampleRate_ = 44100.0;
    float  buffer_[65536] = {};

    inline float processSample(float input, const Parameters& params) {
        // DSP implementation for MB De-Breath
        return input;
    }
};

#endif // MB_VOCAL_DEBREATH_H

File: built-in plugins dsp/mb-vocal-debreath.ts
import type { PluginDefinition } from '../server/services/pluginHostService';

const MbVocalDebreathPlugin: PluginDefinition = { id: 'mb-vocal-debreath', slug: 'mb-vocal-debreath', name: 'MB De-Breath', category: 'effect', type: 'vocal', version: '1.0.0', description: 'Breath noise reduction', author: 'Max Booster', parameters: [{ id: 'sensitivity', name: 'Sensitivity', type: 'float', defaultValue: 0.5, minValue: 0, maxValue: 1, automatable: true }, { id: 'reduction', name: 'Reduction', type: 'float', defaultValue: -12, minValue: -40, maxValue: 0, automatable: true }], defaultPreset: { sensitivity: 0.5, reduction: -12 } };

export default MbVocalDebreathPlugin;
